smithy-trait-codegen/codegen-core/src/main/java/software/amazon/smithy/traitcodegen/generators/common/BuilderGenerator.java

```diff
@@ -0,0 +1,305 @@
+            writeValuesProperty(shape);
+            return null;
+        }
+
+        @Override
+        public Void structureShape(StructureShape shape) {
+            shape.members().forEach(this::writeProperty);
+            return null;
+        }
+
+        private void writeProperty(MemberShape shape) {
+            Optional<String> builderRefOptional =
+                    symbolProvider.toSymbol(shape).getProperty(SymbolProperties.BUILDER_REF_INITIALIZER, String.class);
+            if (builderRefOptional.isPresent()) {
+                writer.addImport(BuilderRef.class);
+                writer.write(BUILDER_REF_TEMPLATE, symbolProvider.toSymbol(shape),
+                        symbolProvider.toMemberName(shape),
+                        builderRefOptional.orElseThrow(RuntimeException::new));
+            } else {
+                writer.write("private $B $L;", symbolProvider.toSymbol(shape),
+                        symbolProvider.toMemberName(shape));
+            }
+        }
+
+        private void writeValuesProperty(Shape shape) {
+            Symbol collectionSymbol = symbolProvider.toSymbol(shape);
+            writer.addImport(BuilderRef.class);
+            writer.write(BUILDER_REF_TEMPLATE, collectionSymbol, VALUES,
+                    collectionSymbol.expectProperty(SymbolProperties.BUILDER_REF_INITIALIZER));
+        }
+    }
+
+
+    private final class BuilderSetterGenerator extends ShapeVisitor.Default<Void> {
+        @Override
+        protected Void getDefault(Shape shape) {
+            throw new UnsupportedOperationException("Does not support shape of type: " + shape.getType());
+        }
+
+        @Override
+        public Void listShape(ListShape shape) {
+            shape.accept(new SetterVisitor(VALUES));
+            return null;
+        }
+
+        @Override
+        public Void mapShape(MapShape shape) {
+            shape.accept(new SetterVisitor(VALUES));
+            return null;
+        }
+
+        @Override
+        public Void structureShape(StructureShape shape) {
+            shape.members().forEach(
+                    memberShape -> memberShape.accept(new SetterVisitor(symbolProvider.toMemberName(memberShape))));
+            writer.newLine();
+            return null;
+        }
+    }
+
+
+    private final class SetterVisitor extends ShapeVisitor.Default<Void> {
+        private final String memberName;
+
+        private SetterVisitor(String memberName) {
+            this.memberName = memberName;
+        }
+
+        @Override
+        protected Void getDefault(Shape shape) {
+            writeStandardAccessors(shape);
+            return null;
+        }
+
+        @Override
+        public Void listShape(ListShape shape) {
+            writeListAccessors(shape);
+            return null;
+        }
+
+        @Override
+        public Void mapShape(MapShape shape) {
+            writeMapAccessors(shape);
+            return null;
+        }
```

Comment:
Why the layer of indirection?

Reply:
No good reason. Ill move the method body here.